url7f43a4752f_shalynnho_pokemonfactory_tgz-pJ8-Networking_ServerJ8/src/agent/LaneAgent.java

```diff
@@ -0,0 +1,320 @@
+		if(state == LaneStatus.PURGING) {
+			state = LaneStatus.BROKEN_WHILE_PURGING;
+		}
+		else { 
+			state = LaneStatus.BROKEN;
+		}
+		stateChanged();
+	}
+	
+	@Override
+	public boolean pickAndExecuteAnAction() {
+		// print("In the Scheduler");
+		
+		if (state == LaneStatus.PURGING || state == LaneStatus.BROKEN_WHILE_PURGING) {
+			purgeSelf();
+			return true;
+		}
+		/*print("state is "+state+" and i have "+requestList.size()+" extra request "+extraRequestCount+" and currently "
+		+currentParts.size());*/
+		
+		synchronized (currentParts) {
+			for (MyPart part : currentParts) {
+				if (part.status == PartStatus.END_LANE) {
+					tellNest(part);
+					return true;
+				}
+			}
+		}
+
+		synchronized (currentParts) {
+			for (MyPart part : currentParts) {
+				if (part.status == PartStatus.NEED_TO_DELIVER) {
+					giveToNest(part);
+					return true;
+				}
+			}
+		}
+		
+		if (state == LaneStatus.WAITING) {
+			if(requestList.size() > lowerThreshold && currentType != null && currentParts.size() != 0) {
+				//extraRequestCount = 0;
+				state = LaneStatus.FILLING;
+				return true;
+			}
+		}
+		
+		if (state == LaneStatus.FILLING && currentType != null) {
+			if(extraRequestCount < topLimit) {
+				requestList.add(currentType);
+				extraRequestCount++;
+				return true;
+			}
+			if(requestList.size() == 0) {
+				//extraRequestCount = 0;
+				state = LaneStatus.WAITING;
+				return true;
+			}
+		}
+		
+		if (state == LaneStatus.FILLING) {
+			synchronized (requestList) {
+				for (PartType requestedType : requestList) {
+					getParts(requestedType);
+					return true;
+				}
+			}
+		}
+		
+		
+		
+		
+		return false;
+	}
+
+	public void purgeSelf() {
+		print("Purging self");
+		requestList = Collections.synchronizedList(new ArrayList<PartType>());
+		currentParts = Collections.synchronizedList(new ArrayList<MyPart>());
+		extraRequestCount = 0;
+		currentType = null;
+		if (laneGUI != null) {
+			laneGUI.purge();
+			try {
+				animation.acquire();
+			} catch (InterruptedException e) {
+				// TODO Auto-generated catch block
+				e.printStackTrace();
+			}
+		}
+		feeder.msgThisLanePurged(this);
+		nest.msgLanePurgeDone();
+		if(state == LaneStatus.BROKEN_WHILE_PURGING) {
+			state = LaneStatus.BROKEN;
+		} else {
+			state = LaneStatus.FILLING;
+		}
+		stateChanged();
+	}
+
+	public void getParts(final PartType requestedType) {
+		print("Telling Feeder that it needs a part");
+		requestList.remove(requestedType);
+		feeder.msgINeedPart(requestedType, this);
+		stateChanged();
+	}
+
+	public void tellNest(MyPart p) {
+		p.status = PartStatus.TOLD_NEST;
+		nest.msgPartReady();
+		stateChanged();
+	}
+
+	public void giveToNest(MyPart mp) {
+		print("Giving part to Nest of type " + mp.part.type.getName());
+		mp.status = PartStatus.DELIVERED;
+		if (laneGUI != null) {
+			laneGUI.givePartToNest(mp.part.partGraphics);
+		}
+		try {
+			animation.acquire();
+		} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)